File: 0202. Happy Number/0202.myans.cpp
class Solution {
public:
    bool isHappy(int n) {
        std::set<int> seen;
        int base = n;
        while (true) {
            if (base == 1) return true;

            auto [it, inserted] = seen.insert(base);
            if (!inserted) return false;

            base = calcSquareSum(base);
        }
    }

    int calcSquareSum(int n) {
        int sum = 0;
        int base = n;
        while (base != 0) {
            int num = base % 10;
            sum += num * num;
            base /= 10;
        }
        return sum;
    }
};

File: 0202. Happy Number/0202.myans.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut base = n;
        let mut seen: HashSet<i32> = HashSet::new();

        loop {
            if base == 1 { return true; }
            if !seen.insert(base) {
                return false;
            } 
            base = Self::calc_square_sum(base);
        }
    }

    fn calc_square_sum(n: i32) -> i32 {
        let mut calc = 0;
        let mut base = n;
        while base != 0 {
            let num = base % 10;
            calc += num * num;
            base /= 10;
        }
        calc
    }
}